src/server/saved_objects/validation/readme.md

```diff
@@ -0,0 +1,63 @@
+# Saved Object Validations
+
+The saved object client supports validation of documents during create / bulkCreate operations.
+
+This allows us tighter control over what documents get written to the saved object index, and helps us keep the index in a healthy state.
+
+## Creating validations
+
+Plugin authors can write their own validations by adding a `validations` property to their uiExports. A validation is nothing more than a dictionary of `{[prop: string]: validationFunction}` where:
+
+* `prop` - a root-property on a saved object document
+* `validationFunction` - a function that takes a document and throws an error if it does not meet expectations.
+
+## Example
+
+```js
+// In myFanciPlugin...
+uiExports: {
+  validations: {
+    myProperty(doc) {
+      if (doc.attributes.someField === undefined) {
+        throw new Error(`Document ${doc.id} did not define "someField"`);
+      }
+    },
+
+    someOtherProp(doc) {
+      if (doc.attributes.counter < 0) {
+        throw new Error(`Document ${doc.id} cannot have a negative counter.`);
+      }
+    },
+  },
+},
+```
+
+In this example, `myFanciPlugin` defines validations for two properties: `myProperty` and `someOtherProp`.
+
+This means that no other plugin can define validations for myProperty or someOtherProp.
+
+The `myProperty` validation would run for any doc that has a `type="myProperty"` or for any doc that has a root-level property of `myProperty`. e.g. it would apply to all documents in the following array:
```

Comment:
I don't love this validation API because it seems unintuitive and requires some knowledge about the underlying storage mechanics of saved objects, but I understand why it exists the way it does for now. I don't think there's a solution here right now without overhauling how we store and treat saved objects, so we can proceed with this. I just wanted to raise the concern.